sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -764,6 +773,184 @@ abstract class TypeCoercionBase {
     }
   }
 
+  object CollationTypeCasts extends TypeCoercionRule {
+    override val transform: PartialFunction[Expression, Expression] = {
+      case e if !e.childrenResolved => e
+
+      case checkCastWithIndeterminate: Concat
+        if shouldCast(checkCastWithIndeterminate.children) =>
+        val newChildren =
+          collateToSingleType(checkCastWithIndeterminate.children, failOnIndeterminate = false)
+        checkCastWithIndeterminate.withNewChildren(newChildren)
+
+      case checkCastWithoutIndeterminate@(_: BinaryExpression | _: In | _: SortOrder)
+        if shouldCast(checkCastWithoutIndeterminate.children) =>
+        val newChildren = collateToSingleType(checkCastWithoutIndeterminate.children)
+        checkCastWithoutIndeterminate.withNewChildren(newChildren)
+
+      case checkIndeterminate@(_: BinaryExpression | _: In | _: SortOrder)
+        if hasIndeterminate(checkIndeterminate.children
+          .filter(e => hasStringType(e.dataType))
+          .map(e => extractStringType(e.dataType))) =>
+        throw QueryCompilationErrors.indeterminateCollationError()
+
+      case checkImplicitCastInputTypes: ImplicitCastInputTypes
+        if checkImplicitCastInputTypes.children.exists(e => hasStringType(e.dataType))
+          && checkImplicitCastInputTypes.inputTypes.nonEmpty =>
+        val collationId: Int =
+          getOutputCollation(checkImplicitCastInputTypes
+            .children.filter { e => hasStringType(e.dataType) })
+        val children: Seq[Expression] = checkImplicitCastInputTypes
+          .children.zip(checkImplicitCastInputTypes.inputTypes).map {
+            case (e, st) if hasStringType(st) =>
+              castStringType(e, collationId, Some(st)).getOrElse(e)
+            case (e, TypeCollection(types)) if types.exists(hasStringType) =>
+              types.flatMap{ dt =>
+                if (hasStringType(dt)) {
+                  castStringType(e, collationId, Some(dt))
+                } else {
+                  implicitCast(e, dt)
+                }
+              }.headOption.getOrElse(e)
+            case (in, expected) => implicitCast(in, expected).getOrElse(in)
+          }
+        checkImplicitCastInputTypes.withNewChildren(children)
+
+      case checkExpectsInputType: ExpectsInputTypes
```

Comment:
I don't think we should handle implicit cast in this new rule. The new rule should only resolve conflicting string collations for string function inputs. We should try our best to not do one thing in multiple places and rely on the rule execution order. It's much easier to reason about if we only have one central place to do implicit cast.

Reply:
I agree and could move implicitCasting from here to `ImpicitTypeCast`. Only problem is that we then have to change multiple checks to run `getOutputCollation` (e.g. `ConcatCoercion`) as they might cast to StringType(0) even though we want collated cast. I was thinking about performance issues if this is ok to run check for collationId multiple times in different checks? Do you have some idea on how heavy this operation would be to the whole casting process, as I am not still quite sure?

Reply:
If all the problems come from `Concat`, shall we require all its inputs to be the same collation in the first version? It's unclear to me how useful is the pgsql semantic for concat.

Reply:
Also, I was a bit sceptical about changing acceptsType in StringType. If I understood correctly this function should give us what types can be used for this expression without implicitCasting. Unfortunately, even after Uros'es PR https://github.com/apache/spark/pull/45422 we will have this problem. We do not want to get a function with two binary collations like UNICODE and UTF8_BINARY, and still say that it is fine to call a function. We specifically need one collation. One workaround would be to just include collationid in the check of all the special cases. This also would change the hierarchy of classes, as additional objects that Uros added would need to extend StringType.

Reply:
Unfortunately, we do not have a trait to indicate the inputs of an expression should all be the same data type. We have various type coercion rules to do it for various expressions: `InConversion`, `ConcatCoercion`, `IfCoercion`, etc.

This is probably not a big problem, as not many functions support more than one string input and we can add a new rule to merge string collations.